Calculate sleep time (go more near the reatelimit)

Aras14HD opened the pull request:
The Reddit API allows 100 requests per minute, we currently do only little more than 3 on average as mentions are rare (one more request then) and most comments do not lead to a reply. To get closer without accidentally going over the limit, I implemented this automatic sleep calculation based on `X-Ratelimit-Remaining` and `X-Ratelimit-Reset`.

During normal operation, it waits a little less on comments than calculated (reset/remaining) to make up for the added delay of get_comments and rapidly replies (every 2 seconds). If it is close to running out, it adds some extra delay and uses a calculated wait for replies. If it runs out (could request too much), it waits a little more than the reset time.

As this would normally recheck comments even more, I used the `after` parameter to only fetch new comments/posts/mentions.

Further:
- [ ] Test and tweak the delay calculation
- [x] Check performance of `after` (whether url query parsing takes less time than checking 100 comments)
- [x] Implement checking posts and mentions less often (every nth loop)

Resolves #49
Resolves #12

Aras14HD commented:
Checked performance, that method only takes ~500ns on my end, so maybe a few micros in deploy. Not a significant slowdown.

tolik518 commented:
Thats pretty neat. Sounds like a pretty good solution 

I can't wait to look into the implementation in the following days

tolik518 commented on README.md:
```diff
@@ -78,14 +78,19 @@ PASSWORD=<reddit_app_password>
-Replace with the values you received from the Reddit App creation. InfluxDB is optional and can be removed if not needed.
+Replace with the values you received from the Reddit App creation.
+InfluxDB is optional and can be removed if not needed.
+The `_EVERY` variables are optional and default to "1".
```
Could you please write a sentence about how the value affects the behavior. Like what would a number higher than 1 mean and what would it mean if one set it to 0.

Aras14HD replied:
0 would panic (division by zero)

tolik518 commented:
Current code always panics. I've set `CHECK_MENTIONS=false` and `CHECK_POSTS=true`

```
2025-03-29 - 23:05:39.211725413 | Polling Reddit for new comments...
thread 'main' panicked at src/reddit_api.rs:255:22:
Failed to extract comments: ParseIntError { kind: InvalidDigit }
```

I'll merge it after the conflicts were resolved.

tolik518 commented on src/reddit_api.rs:
```diff
@@ -294,6 +349,22 @@ impl RedditClient {
 
+        let response_headers = response.headers();
+        let remaining: u64 = response_headers
+            .get("X-Ratelimit-Remaining")
```
It should matter, but `x-ratelimit-remaining` is all smallcaps in the reddit headers

Aras14HD replied:
I used it in caps as that is what the docs write it as.

tolik518 commented on src/reddit_api.rs:
```diff
@@ -294,6 +349,22 @@ impl RedditClient {
+            .to_str()
+            .unwrap()
+            .parse()
+            .unwrap();
```
I'd do here something like that 
```rust
        let remaining: u64 = response_headers
            .get("x-ratelimit-remaining")
            .expect("Missing Ratelimit header")
            .to_str()
            .unwrap()
            .trim()
            .split('.')  // Handle potential decimal point
            .next()
            .unwrap_or("0")
            .parse()?;
```

As I commented, the header sometimes has a decimal point, which leads to panics. Same goes for the x-ratelimit-reset

Aras14HD replied:
Instead of ignoring the decimal, I used floats

tolik518 commented on src/reddit_api.rs:
```diff
@@ -451,10 +522,24 @@ impl RedditClient {
         let empty_vec = Vec::new();
+        let headers = response.headers();
+        let remaining: u64 = headers
+            .get("X-Ratelimit-Remaining")
```
smallcaps and make sure decimals work, as I commented above :)

tolik518 reviewed (changes requested):
I just tested it locally and I'm not sure if it works as it supposed to. The bot didn't reply to my most recent comment, but replied to older ones, which makes me thing `after` check comments that came before the 100 latest comments. 

Also the bot now checks pretty aggressively which uses the ratelimit faster than it would regenerate. Heres a snippet from the logs 

```
2025-03-29 - 23:32:00.013498696 | Polling Reddit for new comments...
Remaining: 695
reset: 479
Remaining: 696
reset: 479
2025-03-29 - 23:32:00.426623935 | Polling Reddit for new comments...
Remaining: 694
reset: 479
Remaining: 693
reset: 479
2025-03-29 - 23:32:01.025071113 | Polling Reddit for new comments...
Remaining: 692
reset: 478
Remaining: 691
reset: 478
2025-03-29 - 23:32:01.511572918 | Polling Reddit for new comments...
Remaining: 690
reset: 478
Remaining: 689
reset: 478
2025-03-29 - 23:32:01.948305637 | Polling Reddit for new comments...
Remaining: 688
reset: 477
Remaining: 687
reset: 477
2025-03-29 - 23:32:02.366859691 | Polling Reddit for new comments...
Remaining: 686
reset: 477
Remaining: 685
reset: 477
2025-03-29 - 23:32:02.782440917 | Polling Reddit for new comments...
Remaining: 684
reset: 477
Remaining: 683
reset: 477
2025-03-29 - 23:32:03.287504642 | Polling Reddit for new comments...
Remaining: 682
reset: 476
Remaining: 681
reset: 476
2025-03-29 - 23:32:03.697507848 | Polling Reddit for new comments...
Remaining: 680
reset: 476
Remaining: 679
reset: 476
2025-03-29 - 23:32:04.111844088 | Polling Reddit for new comments...
Remaining: 677
reset: 475
Remaining: 678
reset: 475
2025-03-29 - 23:32:04.535020506 | Polling Reddit for new comments...
Remaining: 676
reset: 475
Remaining: 675
reset: 475
2025-03-29 - 23:32:04.998443897 | Polling Reddit for new comments...
Remaining: 673
reset: 474
Remaining: 674
reset: 474
2025-03-29 - 23:32:05.414219995 | Polling Reddit for new comments...
Remaining: 671
reset: 474
Remaining: 672
reset: 474
2025-03-29 - 23:32:05.839603484 | Polling Reddit for new comments...
Remaining: 669
reset: 474
Remaining: 670
reset: 474
2025-03-29 - 23:32:06.262308658 | Polling Reddit for new comments...
Remaining: 667
reset: 473
Remaining: 668
reset: 473
```

Aras14HD commented:
The polling behaviour is actually good, staying at a healthy level, it doesn't slowly regenerate, but reset when `reset` hits 0.

The checking old comments problem might be due to the starting value "", which in testing with the posts endpoint for me didn't cause problems, but may have the effect of getting the earliest ones the api is willing to return. That could be solved by not setting the query at all if the value is empty.

Aras14HD commented:
Just wanted to check performance (if there is anything slowing down `get_comments` besides the requests, result: just requests) and noticed that I left in the extraction and calculation of factorials in the "new" method ^_^' To fix that I had to add a better check for possible factorials (used to check if we should check the parent of a mention). That check is just is there a number after or before an `!` or before a `?` or `)` before `!` or `?` or `(` after `!`.

Aras14HD commented:
Now, it doesn't request with `after` on the first request, but only if it has a newest request.

tolik518 commented:
Oh at the current state it's always panicking:

```rust
RUST_BACKTRACE=full cargo run                                                                                                                     101 ✘  took 14s   at 21:08:22  
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/factorion-bot`
InfluxDB client configured. Metrics will be logged.
Fetched new token. Will expire: 2025-04-01T19:09:22Z
Now: 2025-03-31T19:09:22.494967688Z | Expiration time: 2025-04-01T19:09:22Z
Found comment_ids in the file
2025-03-31 - 19:09:22.53747135 | Polling Reddit for new comments...
Comment ID t1_mkiqm2t -> Status { already_replied_or_rejected: false, not_replied: true, number_too_big_to_calculate: false, no_factorial: false, reply_would_be_too_long: false, factorials_found: true } -> [Calculation { value: Int(27), steps: [(1, 1)], result: Exact(-10888869450418352160768000000) }]

thread 'main' panicked at src/calculation_results.rs:294:14:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0:     0x56ce19fdfe10 - std::backtrace_rs::backtrace::libunwind::trace::hda605fcf5cf97597
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x56ce19fdfe10 - std::backtrace_rs::backtrace::trace_unsynchronized::h7b07625bcf26b3f5
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x56ce19fdfe10 - std::sys::backtrace::_print_fmt::hf07723f6e25fc914
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/sys/backtrace.rs:66:9
   3:     0x56ce19fdfe10 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h6dd9c826bc2e3964
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/sys/backtrace.rs:39:26
   4:     0x56ce1a009283 - core::fmt::rt::Argument::fmt::he07b17bdee0d7e43
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/core/src/fmt/rt.rs:177:76
   5:     0x56ce1a009283 - core::fmt::write::h01a5c56b68dfd138
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/core/src/fmt/mod.rs:1449:21
   6:     0x56ce19fdc2b3 - std::io::Write::write_fmt::hdaf8c7fcf6213d53
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/io/mod.rs:1890:15
   7:     0x56ce19fdfc62 - std::sys::backtrace::BacktraceLock::print::h9876829a305e022a
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/sys/backtrace.rs:42:9
   8:     0x56ce19fe0cff - std::panicking::default_hook::{{closure}}::he4cc03e8b4486b03
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panicking.rs:298:22
   9:     0x56ce19fe0b6a - std::panicking::default_hook::hecdf138804f766bd
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panicking.rs:325:9
  10:     0x56ce19fe1722 - std::panicking::rust_panic_with_hook::h5cd781d4400f745e
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panicking.rs:831:13
  11:     0x56ce19fe1476 - std::panicking::begin_panic_handler::{{closure}}::h6e87d44aa0be8310
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panicking.rs:697:13
  12:     0x56ce19fe0319 - std::sys::backtrace::__rust_end_short_backtrace::h292d20b1f5fee88c
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/sys/backtrace.rs:168:18
  13:     0x56ce19fe113d - rust_begin_unwind
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panicking.rs:695:5
  14:     0x56ce1a006820 - core::panicking::panic_fmt::he5be4d23a9ef3d6c
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/core/src/panicking.rs:75:14
  15:     0x56ce1a0068ac - core::panicking::panic::he3ef292a8104537f
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/core/src/panicking.rs:145:5
  16:     0x56ce1a006799 - core::option::unwrap_failed::h45c3d5443b4b4cb8
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/core/src/option.rs:2015:5
  17:     0x56ce1941a891 - core::option::Option<T>::unwrap::hc800c91bc781bc9a
                               at /home/tolik/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:978:21
  18:     0x56ce1941a891 - factorion_bot::calculation_results::Calculation::truncate::h631fb79d8b1628ad
                               at /home/tolik/Projects/factorion-bot-aras/src/calculation_results.rs:292:22
  19:     0x56ce19419063 - factorion_bot::calculation_results::Calculation::format::h66ee10c2e4da927c
                               at /home/tolik/Projects/factorion-bot-aras/src/calculation_results.rs:197:21
  20:     0x56ce194377c1 - factorion_bot::reddit_comment::RedditComment<alloc::vec::Vec<factorion_bot::calculation_results::Calculation>>::get_reply::{{closure}}::h9d23c2864595eeab
                               at /home/tolik/Projects/factorion-bot-aras/src/reddit_comment.rs:582:25
  21:     0x56ce1949cb59 - <core::slice::iter::Iter<T> as core::iter::traits::iterator::Iterator>::fold::hf8ab9da249a15c3c
                               at /home/tolik/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/iter/macros.rs:232:27
  22:     0x56ce19451a16 - factorion_bot::reddit_comment::RedditComment<alloc::vec::Vec<factorion_bot::calculation_results::Calculation>>::get_reply::h177dc56b5322afb8
                               at /home/tolik/Projects/factorion-bot-aras/src/reddit_comment.rs:578:25
  23:     0x56ce194941c5 - factorion_bot::main::{{closure}}::ha4eb95497c7ebad8
                               at /home/tolik/Projects/factorion-bot-aras/src/main.rs:174:37
  24:     0x56ce1943420b - <core::pin::Pin<P> as core::future::future::Future>::poll::h72d6a5287581a4d7
                               at /home/tolik/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/future.rs:124:9
  25:     0x56ce1948c68e - tokio::runtime::park::CachedParkThread::block_on::{{closure}}::h424d6121ac125e0c
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/runtime/park.rs:284:60
  26:     0x56ce1948b736 - tokio::task::coop::with_budget::h3eb1c8fd24f7ee91
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/task/coop/mod.rs:167:5
  27:     0x56ce1948b736 - tokio::task::coop::budget::h78751eae0a6574ae
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/task/coop/mod.rs:133:5
  28:     0x56ce1948b736 - tokio::runtime::park::CachedParkThread::block_on::h082fefb6410e470b
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/runtime/park.rs:284:31
  29:     0x56ce19422bcf - tokio::runtime::context::blocking::BlockingRegionGuard::block_on::h22d1e7a9adfeb667
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/runtime/context/blocking.rs:66:9
  30:     0x56ce19452800 - tokio::runtime::scheduler::multi_thread::MultiThread::block_on::{{closure}}::h63dc14831b4ead9a
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/runtime/scheduler/multi_thread/mod.rs:87:13
  31:     0x56ce19416a13 - tokio::runtime::context::runtime::enter_runtime::ha69ff12844a34527
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/runtime/context/runtime.rs:65:16
  32:     0x56ce19452761 - tokio::runtime::scheduler::multi_thread::MultiThread::block_on::h5a3a5a87d4ad46dd
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/runtime/scheduler/multi_thread/mod.rs:86:9
  33:     0x56ce19445cb8 - tokio::runtime::runtime::Runtime::block_on_inner::h1f98ad120db57cf8
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/runtime/runtime.rs:370:45
  34:     0x56ce19445fdb - tokio::runtime::runtime::Runtime::block_on::h78f719c2c72eccbb
                               at /home/tolik/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.1/src/runtime/runtime.rs:340:13
  35:     0x56ce194962ee - factorion_bot::main::he7f0cf2deeca6d88
                               at /home/tolik/Projects/factorion-bot-aras/src/main.rs:105:5
  36:     0x56ce19457d1b - core::ops::function::FnOnce::call_once::h87be578b96004061
                               at /home/tolik/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  37:     0x56ce1945631e - std::sys::backtrace::__rust_begin_short_backtrace::h09049c9e6c98345e
                               at /home/tolik/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/backtrace.rs:152:18
  38:     0x56ce1940b671 - std::rt::lang_start::{{closure}}::h0aeefa9e75740c59
                               at /home/tolik/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:199:18
  39:     0x56ce19fd5a70 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hf8d804b471050a2c
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/core/src/ops/function.rs:284:13
  40:     0x56ce19fd5a70 - std::panicking::try::do_call::heb46fd7038a5c631
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panicking.rs:587:40
  41:     0x56ce19fd5a70 - std::panicking::try::h1ab07475f01f8370
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panicking.rs:550:19
  42:     0x56ce19fd5a70 - std::panic::catch_unwind::h493f8c2484e4d941
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panic.rs:358:14
  43:     0x56ce19fd5a70 - std::rt::lang_start_internal::{{closure}}::h160a9a283f927f22
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/rt.rs:168:24
  44:     0x56ce19fd5a70 - std::panicking::try::do_call::h41bfcae2a8e350d7
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panicking.rs:587:40
  45:     0x56ce19fd5a70 - std::panicking::try::h9e2b0c5a3f42bad1
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panicking.rs:550:19
  46:     0x56ce19fd5a70 - std::panic::catch_unwind::h3309ccff3e64bfa0
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/panic.rs:358:14
  47:     0x56ce19fd5a70 - std::rt::lang_start_internal::h44c07f550f530434
                               at /rustc/6171d944aea415a3023d4262e0895aa3b18c771f/library/std/src/rt.rs:164:5
  48:     0x56ce1940b657 - std::rt::lang_start::h1ea8172adc98d74b
                               at /home/tolik/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:198:5
  49:     0x56ce194963be - main
  50:     0x7d89ee629d90 - <unknown>
  51:     0x7d89ee629e40 - __libc_start_main
  52:     0x56ce19409465 - _start
  53:                0x0 - <unknown>
```

Aras14HD commented:
That problem happens in the truncation after turning the number into a float and then calculating the ln of that. (From an exact number result in this case) So there are two possibilities:
- the number was negative (most likely reason and easy to fix)
- the number was too large (unlikely for a number around 10^1000000, but I'll have to check again)

Aras14HD commented:
Just rechecked with the "after" checking, the `after` query does not give you stuff after that point in time, but rather afterwards in the listing. Switched to before, should work correctly.

Aras14HD commented:
I checked, if it would merge, all conflicts (with the other pulls) are trivial to resolve, just version and for some reason in math there's a conflict with the ratelimit side empty (just merge the actual change).

Aras14HD commented:
I finally set it up, so I can test it on Reddit (using Aras14Alt), added the option to not reply for that and fixed the out of date readme documentation. Also fixed a bug when SUBREDDITS is empty. Otherwise, on my end, it is running fine.

tolik518 commented:
Oh perfect, thank you for updating the docs! I didn't realize it was out of date.

I'll  merge the other PRs first, then I'll get to that one.

Aras14HD commented:
So with testing, I am currently making some adjustments, will push another commit soon

tolik518 commented:
Okay, tag we again when it's ready :)

Aras14HD commented:
@tolik518 pushed that commit, should now always run a little slower than the perfect rate, so it can recover from large comment bursts. (better steady than fast)

If you don't find any problems, it should be ready now :)

tolik518 commented:
I'll check it out now :)

(Un)related question: Do you update the factorion version in the .lock file by hand, or do you run a command for that? Because I haven't actively updated it in the lock-file

Aras14HD commented:
I update it, when I merge, after that (if I bump), it normally happens automatically or latest at the next test.

tolik518 commented:
I skimmed over the changes and haven't seen anything obvious. Also locally it runs as expected, so I'll merge it now!

Also I like the new pacing a lot better, it feels less risky to get unexpectedly to the limit :D

tolik518 commented:
FYI: I set the new variables to the following values:
```ini 
CHECK_MENTIONS=true
CHECK_POSTS=true
MENTIONS_EVERY=2
POSTS_EVERY=2
```

Also I just realized that the `SLEEP_BETWEEN_REQUESTS` is now unused and can be removed next time anyone touches the repo

Aras14HD commented:
So the api is returning an empty error, that is not much to go off. However looking at `reply_to_comment` I notice, that it doesn't check the jwt expiry. Another cause could be, that even the 3/2 seconds are too fast for reddit, and it is rate-limiting you? (Would mean to increase the wait time before replies, might have a stricter one for that)

Aras14HD commented:
And on mentions, it has missed some previously, if they were in reply, like [here](https://www.reddit.com/r/mathematics/s/pfP03CnLYf). Maybe they didn't show up in the mentions listing or in a different order. Could be something that Reddit does if you just mentioned him.
